src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+// Written in the D programming language.
+/**
+This module provides a composable reference count implementation in the form
+of `__RefCount`.
+
+License: $(HTTP www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+
+Authors: Eduard Staniloiu
+
+Source: $(DRUNTIMESRC core/experimental/refcount.d)
+*/
+module core.experimental.refcount;
+
+///
+unittest
+{
+    auto rc = __RefCount.make!__RefCount();
+    assert(rc.isUnique);
+    {
+        auto rc2 = rc; // Construct rc2 by copy construction
+        assert(!rc.isUnique);
+        auto rc3 = __RefCount.make!__RefCount();
+        rc2 = rc3; // Assign rc3 into rc2; rc's ref count drops
+        assert(rc.isUnique);
+        rc3 = rc; // Assign rc into rc3; rc's ref count increases
+        assert(!rc.isUnique);
+        // rc2 and rc3 go out of scope here
+        // rc2 is the last ref to the `__RefCount` created in this block -> gets freed
+    }
+    assert(rc.isUnique);
+}
+
+/**
+A qualified reference counted `struct` that is intended to be composed by
+user-defined types that desire to implement manual memory management by means of
+reference counting.
+
+`__RefCount` was designed to be composed as a field inside a user-defined type.
+The user is responsible to initialize the `__RefCount` in that type's constructors.
+The user will call the `isUnique()` method to decide if this is the last
+reference to the enclosing type so memory can be safely deallocated.
+
+Implementation: The internal implementation of `__RefCount` uses `malloc`/`free`.
+
+Important: The `__RefCount` member must be initialized through a call to its
```

Comment:
Can't we just disable the default initialization to enforce this?

```d
@disable this();
```

Reply:
This breaks `rcarray` for some reason: ```src/core/experimental/array.d(676): Error: variable `core.experimental.array.rcarray!int.rcarray.opSlice!(immutable(rcarray!int)).opSlice.__copytmp2437` default construction is disabled for type immutable(__RefCount)```

Reply:
Sorry, where is `src/core/experimental/array.d`? somewhere in your fork of druntime?